### PR TITLE
Dev extend spi interface

### LINKED ISSUE #301: Extend SPI Interface
SPI Interface lacks simple Write and Read operations. This shall be added to the interface definition.

### PR DESCRIPTION
**Pull request by c0nwild**
This PR extends SPI capabilities:
- Simple Write operation
- Chip select pin active state is configurable

Acceptance criteria:
- Unit tests cover 100% and all pass
- Unit tests are complete (no logical fallacy is untested and all equivalency classes were tested)
- SPI class diagram from wiki is not in conflict with implementation
- All automatic checks pass
- Google Styleguide was followed
- Additional guidelines from the wiki were followed
- No logical errors in code
- All error return values are always checked

Closes #301

### CODE REVIEW

**Comment by elheck** [comment]
@ThorsHamster Do you know why this PR doesn't automatically merge?

**Comment by ThorsHamster** [comment]
@Mergifyio refresh
Refresh Mergify after update of Rules.

**Comment by ThorsHamster** [comment]
@Mergifyio refresh

File: src/spi/spi.cpp [hunk 1]
```diff
@@ -1,6 +1,37 @@
 #include "spi.hpp"
 
 namespace spi {
+auto SPI::Write(std::vector<uint8_t> &mosi_data_buffer) noexcept -> types::DriverStatus {
+  HAL_StatusTypeDef transmit_ret_value = HAL_ERROR;
+  types::DriverStatus return_value = types::DriverStatus::HAL_ERROR;
+
+  uint8_t transaction_length = static_cast<uint8_t>(mosi_data_buffer.size());
```
Comment (ThorsHamster) [inline src/spi/spi.cpp]: I think it would be more c++ like to use std::uint8_t instead of global uint8_t.

File: src/spi/spi.cpp [hunk 2]
```diff
@@ -1,6 +1,37 @@
...
+
+  if (transmit_ret_value == HAL_OK) {
+    return_value = types::DriverStatus::OK;
+  }
+  if (transmit_ret_value == HAL_TIMEOUT) {
+    return_value = types::DriverStatus::TIMEOUT;
+  }
+  if (transmit_ret_value == HAL_ERROR) {
+    return_value = types::DriverStatus::HAL_ERROR;
+  }
```
Comment (ThorsHamster) [inline src/spi/spi.cpp]: You use this construct two times. Could also be one method to remove duplication

File: tests/spi/spi_tests.cpp [hunk 3]
```diff
@@ -10,62 +10,122 @@ class SPITests : public ::testing::Test {
...
 
-TEST_F(SPITests, spi_hal_busy) {
+TEST_F(SPITests, transfer_spi_hal_busy) {
   hspi1.mock_return_value = HAL_StatusTypeDef::HAL_BUSY;
   types::DriverStatus rv = unit_under_test_->Transfer(miso_buffer, mosi_buffer);
   EXPECT_EQ(rv, types::DriverStatus::HAL_ERROR);
 }
 
+TEST_F(SPITests, write_exceeding_buffer_size_limit) {
+  std::vector<uint8_t> too_large_buffer(65);
```
Comment (ThorsHamster) [inline tests/spi/spi_tests.cpp]: Here also should everywhere used std::uint8_t

**Review by ThorsHamster — COMMENTED** [review]
LGTM
Just some minor findings. Good jobs!
Readability is much better now!

File: src/spi/spi.cpp (hunk 1 above)
Comment (c0nwild) [reply src/spi/spi.cpp]: Fixed

File: src/spi/spi.cpp (hunk 2 above)
Comment (c0nwild) [reply src/spi/spi.cpp]: Factored out

File: tests/spi/spi_tests.cpp (hunk 3 above)
Comment (c0nwild) [reply tests/spi/spi_tests.cpp]: fixed

**Review by ThorsHamster — APPROVED** [review]
LGTM
Good Job!

**Pull request merged**